man/log.3
.\"   Part of publib.
.\"  
.\"   Copyright (c) 1994-2006 Lars Wirzenius.  All rights reserved.
.\"  
.\"   Redistribution and use in source and binary forms, with or without
.\"   modification, are permitted provided that the following conditions
.\"   are met:
.\"  
.\"   1. Redistributions of source code must retain the above copyright
.\"      notice, this list of conditions and the following disclaimer.
.\"  
.\"   2. Redistributions in binary form must reproduce the above
.\"      copyright notice, this list of conditions and the following
.\"      disclaimer in the documentation and/or other materials provided
.\"      with the distribution.
.\"  
.\"   THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
.\"   OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\"   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\"   ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
.\"   DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\"   GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\"   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\"   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\"   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\"   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"  
.\" part of publib
.\" "@(#)publib-framework:$Id: log.3,v 1.4 1997/05/09 13:49:05 liw Exp $"
.\"
.TH LOG 3 "C Programmer's Manual" Publib "C Programmer's Manual"
.SH NAME
log \- simple log file output
.SH SYNOPSIS
.nf
#include <publib.h>
.sp 1
int \fBlog_open\fR(const char *\fIfilename\fR, int \fImin_level\fR);
int \fBlog_add\fR(FILE *\fIf\fR, int \fImin_level\fR);
void \fBlog_set_level\fR(int \fIlog_handle\fR, int \fImin_level\fR);
void \fBlog_set_localtime\fR(int \fIlog_handle\fR, int \fIuse_localtime\fR);
int \fBlog_close\fR(void);
void \fBlog_off\fR(void);
void \fBlog_on\fR(void);
void \fBlog_chat\fR(const char *\fIfmt\fR, ...);
void \fBlog_note\fR(const char *\fIfmt\fR, ...);
void \fBlog_warn\fR(const char *\fIfmt\fR, ...);
void \fBlog_error\fR(const char *\fIfmt\fR, ...);
void \fBlog_fatal\fR(const char *\fIfmt\fR, ...);
void \fBlog_printf\fR(int \fIlevel\fR, const char *\fIfmt\fR, ...);
.SH DESCRIPTION
These routines implement a simple log file facility.
There can be several log files open at the same time
(up to 256, but limits on the
number of concurrently open files may set a lower limit).
Each log message is associated with a level of severity,
which can be any integer, but the levels 
\fIlog_level_chat\fR,
\fIlog_level_note\fR,
\fIlog_level_warn\fR,
\fIlog_level_error\fR, and
\fIlog_level_fatal\fR
are predefined, and can easily be used by calling the respective
print functions (\fIlog_chat\fR, etc.); for other log levels,
\fIlog_printf\fR must be called.
Each open log file is associated by a minimum level.
Messages below the minimum level are not output to that log file.
.PP
The log printing functions attempt to force the messages to
be written to the disk (or wherever they're going), using
\fIfflush\fR(3).
This will hopefully avoid problems due to delayed messages
that buffering may cause.
.PP
Log files are opened with \fIlog_open\fR.  An already open file
(such as \fIstderr\fR) can be added to the list of open log files
with \fIlog_add\fR.  (Note that the file should be opened in append
mode, "a" to \fIfopen\fR(3)).
Both of these functions set the minimum level as well.
They return a log file handle, which is used to identify
the log file to the other functions.
The minimum level can later be adjusted with \fIlog_set_level\fR.
\fIlog_close\fR closes all log files.
.PP
By default, times in log files are reported in UTC,
but each log file can be set to use local time with
\fIlog_set_localtime\fR.
.PP
Logging may temporarily be turned off for all log files at the
same time with \fIlog_off\fR.
It can be turned back on with \fIlog_on\fR.
.SH "RETURN VALUE"
\fIlog_open\fR and \fIlog_add\fR return a log file handle.
\fIlog_close\fR returns -1 if the closing of any log file failed,
0 if all went well.
.SH "SEE ALSO"
publib(3), syslog(2)
.SH AUTHOR
Lars Wirzenius (lars.wirzenius@helsinki.fi)


NEWS
NEWS about the publib project
=============================

Version 0.40
------------

    * Lars Wirzenius has passed maintenance of publib to
      Antti-Juhani Kaijanaho.

    * Publib no longer includes the framework; instead, a conventional
      configure--make build system is provided.  Publib now consists
      of the library modules formerly known as liw-modules.

    * Minor bugfixes have been made to the library modules.


Version 0.39
------------

    * liw/cfgtool has been removed. It was obsolete since about a decade.
    
    * liw/strutils/strsplit has been fixed, thanks to bug report from
      Ron Fox.


log/log.c
/* Part of publib.

   Copyright (c) 1994-2006 Lars Wirzenius.  All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:

   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

   THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS
   OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
   DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
   GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
/*
 * log.c -- log files
 *
 * Part of publib.  See man page for more information
 * "@(#)publib-log:$Id: log.c,v 1.12 1997/05/09 13:49:16 liw Exp $"
 */

#include <assert.h>
#include <time.h>
#include <unistd.h>
#include "publib/log.h"
#include "publib/errormsg.h"


#define MAX	256
static struct {
	FILE *f;
	int min;
	int use_localtime;
} logs[MAX];
static int nlogs = 0;
static int log_is_on = 1;

int log_open(const char *filename, int min_level) {
	FILE *f;

	assert(filename != NULL);

	assert(nlogs >= 0 && nlogs <= MAX);
	if (nlogs == MAX) {
		__publib_error("too many log files");
		return -1;
	}

	f = fopen(filename, "a");
	if (f == NULL) {
		__publib_error("fopen failed");
		return -1;
	}

	logs[nlogs].f = f;
	logs[nlogs].min = min_level;
	logs[nlogs].use_localtime = 0;
	return nlogs++;
}

int log_add(FILE *f, int min_level) {
	assert(f != NULL);

	assert(nlogs >= 0 && nlogs <= MAX);
	if (nlogs == MAX) {
		__publib_error("too many log files");
		return -1;
	}

	logs[nlogs].f = f;
	logs[nlogs].min = min_level;
	logs[nlogs].use_localtime = 0;
	return nlogs++;
}

void log_set_level(int logid, int min_level) {
	assert(logid >= 0);
	assert(logid < nlogs);
	logs[logid].min = min_level;
}


void log_set_localtime(int logid, int use_localtime) {
	assert(logid >= 0);
	assert(logid < nlogs);
	logs[logid].use_localtime = use_localtime;
}


int log_close(void) {
	int i, ret;

	assert(nlogs >= 0 && nlogs <= MAX);
	ret = 0;
	for (i = 0; i < nlogs; ++i) {
		if (fclose(logs[i].f) != 0) {
			__publib_error("fclose failed for logfile");
			ret = -1;
		}
	}
	return ret;
}


void log_off(void) {
	log_is_on = 0;
}


void log_on(void) {
	log_is_on = 1;
}


/* All the functions log_chat through log_fatal are exactly the same,
 * except for the level argument to log_vprintf.  Therefore, we use
 * a macro to emphasize the similarity and to avoid unnecessary
 * differences.
 */
#define do_log_at_level(level) \
	void log_##level(const char *fmt, ...) { \
		va_list args; \
		struct tm *tm; \
		time_t t; \
		char buf1[100]; \
		char buf2[200]; \
		const char *p; \
		int i; \
		long pid; \
		if (!log_is_on) return; \
		p = get_progname(); \
		if (*p == '\0') p = "unknown"; \
		tm = NULL; \
		pid = (long) getpid(); \
		for (i = 0; i < nlogs; ++i) { \
			if (log_level_##level < logs[i].min) continue; \
			if (tm == NULL) { \
				time(&t); \
				if (logs[i].use_localtime) \
					tm = localtime(&t); \
				else \
					tm = gmtime(&t); \
				strftime(buf1, sizeof(buf1), \
				"%Y-%m-%d %H-%M-%S %Z %%.50s %%ld %%d ", tm); \
				sprintf(buf2, buf1, p, pid, log_level_##level);\
			} \
			va_start(args, fmt); \
			(void) fputs(buf2, logs[i].f); \
			(void) vfprintf(logs[i].f, fmt, args); \
			(void) fflush(logs[i].f); \
			va_end(args); \
		} \
	}
do_log_at_level(chat)
do_log_at_level(note)
do_log_at_level(warn)
do_log_at_level(error)
do_log_at_level(fatal)


void log_printf(int level, const char *fmt, ...) {
	va_list args;
	struct tm *tm;
	time_t t;
	char buf1[100];
	char buf2[100];
	const char *p;
	int i;
	long pid;

	if (!log_is_on)
		return;

	p = get_progname();
	if (*p == '\0')
		p = "unknown";
	tm = NULL;
	pid = (long) getpid();

	for (i = 0; i < nlogs; ++i) {
		if (level >= logs[i].min) {
			if (tm == NULL) {
				time(&t);
				if (logs[i].use_localtime)
					tm = localtime(&t);
				else
					tm = gmtime(&t);
				strftime(buf1, sizeof(buf1),
				"%Y-%m-%d %H-%M-%S %Z %%.50s %%ld %%d ", tm);
				sprintf(buf2, buf1, p, pid, level);
			}
			va_start(args, fmt);
			(void) fputs(buf2, logs[i].f);
			(void) vfprintf(logs[i].f, fmt, args);
			(void) fflush(logs[i].f);
			va_end(args);
		}
	}
}
